### DISCUSSION TITLE
Is there a way to group by / reduce.

### DISCUSSION THREAD

**Discussion by JAForbes** [open-policy-agent/community#366] [Community]
I've got a list of bit flags, I'd like to reduce them down to a single flag with `bit_or`.  I can't figure out how to perform the reduce within rego.  Any tips would be great!

**Comment by srenatus** [comment]
Heya! It's actually a bit of a hard problem to do generic group_by/reduce in Rego.

But as for workarounds, I'd try splitting the bit flags into number arrays, collect all the number arrays in another array, say `xs`, and try
```rego
n(i) := 1 if {
  some x in xs
  x[i] == 1
} else := 0

solution := join([ n(i) | i := numbers.range(1, count(xs))])

join(arr) := ... # left as an exercise ;)
```

**Reply by anderseknert** [reply]
A generic way of doing reducde is is something I've been wanting for quite some time. Your question had me finally get around to create an issue for that :) https://github.com/open-policy-agent/opa/issues/5779

**Reply by anderseknert** [reply]
If you don't mind me being nosy @JAForbes, I would love to understand more about the policy you're working on! And what domain it applies to.

Sounds intriguing! 😃

**Reply by JAForbes** [reply]
Ahhh a few projects actually.  I'm still getting my head around the syntax, but given the WASM support I was thinking rego could be used as a shared model for form/endpoint validation across various runtimes.

But for this problem, I was just trying to model a permission system I designed at another company to get my head around the concepts.  At that company, each permission has a bit flag, you bit_or together to resolve their capabilities as a single integer.

**Reply by anderseknert** [reply]
Gotcha! Yeah, I could definitely see that being used in older systems. Might not be the best way of approaching the problem today / in Rego, but having something like reduce to allow a 1:1 mapping would be interesting nonetheless.

**Reply by JAForbes** [reply]
yeah eventually you run out of flags...

yeah I think it would be useful generally, very interested to see how this turns out

**Comment by anderseknert** [comment]
As for group-by: https://www.openpolicyagent.org/docs/latest/policy-performance/#comprehension-indexing
